File: java/variable-length-quantity/src/main/java/VariableLengthQuantity.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

class VariableLengthQuantity {
	List<String> encode(List<Long> numbers) {
		return numbers.stream().flatMap(number -> encodeOneNumber(number).stream()).collect(Collectors.toList());
	}

	List<String> encodeOneNumber(long number) {
		List<String> result = new ArrayList<String>();
		String s = Long.toBinaryString(number);
		for (int i = 0; i < s.length(); i += 7) {
			result.add(0,
					"0x" + Integer.toHexString(
							Integer.parseInt(s.substring(Math.max(0, s.length() - i - 7), s.length() - i), 2)
									+ ((i == 0) ? 0 : (1 << 7))));
		}
		return result;
	}

	List<String> decode(List<Long> bytes) {
		if (bytes.get(bytes.size() - 1) >= 1 << 7) {
			throw new IllegalArgumentException("Invalid variable-length quantity encoding");
		}

		List<String> result = new ArrayList<String>();
		int beginIndex = 0;
		for (int i = 0; i < bytes.size(); i++) {
			if (bytes.get(i) < 1 << 7) {
				result.add(decodeOneNumber(bytes.subList(beginIndex, i + 1)));
				beginIndex = i + 1;
			}
		}
		return result;
	}

	String decodeOneNumber(List<Long> bytes) {
		return "0x" + Long.toHexString(Long.parseLong(String.join("",
				bytes.stream().map(b -> String.format("%7s", Long.toBinaryString(b % (1 << 7))).replace(' ', '0'))
						.collect(Collectors.toList())),
				2));
	}
}

File: python/variable-length-quantity/variable_length_quantity.py
def encode(numbers):
    return sum(map(encode_one_number, numbers), [])


def encode_one_number(number):
    s = bin(number)[2:]
    return [int(s[max(0, len(s) - i - 7):len(s) - i], 2) + (0 if i == 0 else 1 << 7) for i in range(0, len(s), 7)][::-1]


def decode(bytes_):
    if bytes_[-1] >= 1 << 7:
        raise ValueError('Incomplete sequence')

    result = []
    begin_index = 0
    for i in range(len(bytes_)):
        if bytes_[i] < 1 << 7:
            result.append(decode_one_number(bytes_[begin_index:i + 1]))
            begin_index = i + 1
    return result


def decode_one_number(bytes_):
    return int(''.join(map(lambda b: '{:0>7}'.format(bin(b % (1 << 7))[2:]), bytes_)), 2)